programs/marinade-finance/src/instructions/management/redelegate.rs

```diff
@@ -0,0 +1,394 @@
+        if source_validator.active_balance
+            <= source_validator_stake_target + self.state.stake_system.min_stake
+        {
+            msg!(
+                "Source validator {} stake {} is <= target {} +min_stake",
+                source_validator.validator_account,
+                source_validator.active_balance,
+                source_validator_stake_target
+            );
+            // return rent for unused accounts
+            self.return_rent_unused_stake_account(self.split_stake_account.to_account_info())?;
+            self.return_rent_unused_stake_account(self.redelegate_stake_account.to_account_info())?;
+            return Ok(()); // Not an error. Don't fail other instructions in tx
+        }
+
+        // compute how much we can unstake from this validator, and cap redelegate amount to it
+        // can't remove more than needed to reach target, and can't remove more than what's in the account
+        let max_redelegate_from_source_account = min(
+            source_validator.active_balance - source_validator_stake_target,
+            stake.last_update_delegated_lamports,
+        );
+
+        // get dest validator from index
+        let mut dest_validator = self.state.validator_system.get_checked(
+            &self.validator_list.data.as_ref().borrow(),
+            dest_validator_index,
+            &self.dest_validator_account.key(),
+        )?;
+
+        // compute dest validator target
+        let dest_validator_stake_target = self
+            .state
+            .validator_system
+            .validator_stake_target(&dest_validator, total_stake_target)?;
+        // verify: dest validator must be under target
+        if dest_validator.active_balance + self.state.stake_system.min_stake
+            >= dest_validator_stake_target
+        {
+            msg!(
+                "Dest validator {} stake+min_stake {} is > target {}",
+                dest_validator.validator_account,
+                dest_validator.active_balance,
+                dest_validator_stake_target
+            );
+            // return rent for unused accounts
+            self.return_rent_unused_stake_account(self.split_stake_account.to_account_info())?;
+            self.return_rent_unused_stake_account(self.redelegate_stake_account.to_account_info())?;
+            return Ok(()); // Not an error. Don't fail other instructions in tx
+        }
+
+        // compute how much we can stake in dest-validator, and cap redelegate amount to it
+        // can't send more than needed to reach target
+        let max_space_dest_validator = dest_validator_stake_target - dest_validator.active_balance;
+        let redelegate_amount_theoretical =
+            min(max_space_dest_validator, max_redelegate_from_source_account);
+
+        // compute how much this particular account will have after split
+        // do not use saturating_sub because an underflow here means bug and should panic
```

Comment:
Just to let you know, you are correct here. But I don't think readers will be happy to dive so deep into checked/overflow_check holly war instead of trying to understand what is going on in this code